enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java

```diff
@@ -791,10 +796,75 @@ private void markAsWaitingFor( ForsetiLockManager.Lock lock, ResourceType type,
             // after we've generated a description of it.
             if ( lock.detectDeadlock( id() ) != -1 )
             {
-                waitList.clear();
-                throw new DeadlockDetectedException( message );
+                // If the deadlock is real, then an owner of this lock must be (transitively) waiting on a lock that
+                // we own. So to verify the deadlock, we traverse the lock owners and their `waitingForLock` fields,
+                // to find a lock that has us among the owners.
+                // We only act upon the result of this method if the `tries` count is above some threshold. The reason
+                // is that the Lock.collectOwners, which is algorithm relies upon, is inherently racy, and so only
+                // reduces the probably of a false positive, but does not eliminate them.
+                if ( isDeadlockReal( lock, tries ) )
+                {
+                    // After checking several times, this really does look like a real deadlock.
+                    waitList.clear();
+                    waitingForLock = null;
+                    throw new DeadlockDetectedException( message );
+                }
             }
         }
+        waitingForLock = null;
+    }
+
+    private boolean isDeadlockReal( ForsetiLockManager.Lock lock, int tries )
+    {
+        Set<ForsetiLockManager.Lock> waitedUpon = new HashSet<>();
```

Comment:
Agreed. This prompted a bunch of reading PG source code, and recognizing two interesting things: One; PG will block on a lock a solid second before it even tries detecting a deadlock; the fact that that does not appear to be an issue corroborates anecdotal experience that real-world deadlocks are really rare; it is likely not much of a performance penalty if detection is slow.

More interestingly though, the implementation contains a bunch of code for reordering wait lists on deadlock; some deadlocks can be removed by doing so:

Imagine two locks, `{l1, l2}`, and three clients `{A, B, C}`:

    (l1 sharedHolders=[A], waitList=[B(exclusive)])
    (l2 exclusive=[C], waitList=[A(exclusive)])

Ie: 
    A has a shared lock on l1 and waits to get a shared lock on l2
    B holds no locks, and is waiting to get an exclusive lock on l1
    C holds a shared lock on l1, blocking A

If C now decides it wants a shared lock on l1, it needs to wait behind
B; that will cause a deadlock. Forseti will simply abort one of them.

However, a better solution would be to reorder the wait list; if C is 
allowed to move ahead of B, it can be granted the shared lock on l1, and
the deadlock disappears.